src/ck-core/ckrdma.C

```diff
@@ -2121,3 +2130,245 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+                cudaMemcpyDeviceToDevice, postStructs[i].cuda_stream));
+          break;
+        }
+      case CkNcpyModeDevice::IPC:
+        {
+          cuda_ipc_device_info& device_info = CsvAccess(gpu_manager).cuda_ipc_device_infos[source.device_idx];
+
+          // 1. Make user-provided stream wait for IPC event using cudaStreamWaitEvent
+          //    (source buffer to device comm buffer on source)
+          hapiCheck(cudaStreamWaitEvent(postStructs[i].cuda_stream, device_info.src_event_pool[source.event_idx], 0));
+
+          // 2. Invoke cudaMemcpyAsync (from source device comm buffer to destination buffer)
+          hapiCheck(cudaMemcpyAsync((void*)dest.ptr, (void*)((char*)device_info.buffer + source.comm_offset),
+                std::min(source.cnt, dest.cnt), cudaMemcpyDeviceToDevice, postStructs[i].cuda_stream));
+
+          // 3. Record IPC event so that the sender can query it for freeing
+          //    device comm buffer and corresponding pair of CUDA IPC events
+          hapiCheck(cudaEventRecord(device_info.dst_event_pool[source.event_idx], postStructs[i].cuda_stream));
+
+          // 4. Set flag in shared memory so that the sender can start querying
+          //    completion of the IPC event
+          cuda_ipc_event_shared* shm_event_shared = (cuda_ipc_event_shared*)((char*)CsvAccess(gpu_manager).shm_ptr
+                + CsvAccess(gpu_manager).shm_chunk_size * source.device_idx
+                + sizeof(cudaIpcMemHandle_t)) + source.event_idx;
+          pthread_mutex_lock(&shm_event_shared->lock);
+          shm_event_shared->dst_flag = true;
+          pthread_mutex_unlock(&shm_event_shared->lock);
+          break;
+        }
+      case CkNcpyModeDevice::RDMA:
+        {
+          // Transfer the received/unpacked data on host to the destination device buffer
+          // TODO: Use GPUDirect RDMA
+          hapiCheck(cudaMemcpyAsync((void*)dest.ptr, source.data, std::min(source.cnt, dest.cnt),
+                cudaMemcpyHostToDevice, postStructs[i].cuda_stream));
+          break;
+        }
+      default:
+        {
+          CkAbort("Invalid mode");
+          break;
+        }
+    }
+
+    // Add source callback for polling, so that it can be invoked once the transfer is complete
+    CkCallback* cb = new CkCallback(source.cb);
+    hapiAddCallback(postStructs[i].cuda_stream, cb);
+  }
+
+  return is_inline;
+}
+
+// Unused, left for future reference
+/*
+int CkRdmaGetDestPEChare(int dest_pe, void* obj_ptr) {
+  // Mechanism extracted from _prepareMsg() in ck.C
+  if (dest_pe < 0) {
+    int pe = -(dest_pe+1);
+    if (pe == CkMyPe()) {
+      VidBlock* vblk = CkpvAccess(vidblocks)[(CmiIntPtr)obj_ptr];
+      void *objPtr = vblk->getLocalChare();
+      dest_pe = objPtr ? pe : vblk->getActualID().onPE;
+    } else {
+      dest_pe = pe;
+    }
+  }
+
+  return dest_pe;
+}
+*/
+
+static void findFreeIpcEvents(DeviceManager* dm, const size_t comm_offset, int& event_idx) {
```

Comment:
Is there a particular reason you're not returning `event_idx` using the function, instead of having it be written using the reference in the function parameter list?

Reply:
I don't know why I did that, will update